### ISSUE TITLE
use a URI type from a library instead of plain Strings

### ISSUE THREAD

**Issue by Be-ing** [bilelmoussaoui/ashpd#69]
The D-Bus API makes use of URIs in several places, for example FileChooser. ashpd currently treats these URIs as `std::string::String`s. It could be helpful to downstream developers to use a URI type from another crate for these URIs instead of simple Strings. I ran into an issue where I was [not properly handling URI encoding](https://codeberg.org/Be.ing/moire/issues/4) which would have been easier to prevent if ashpd encoded more information about the data from the D-Bus API than treating it as a generic String.

**Comment by bilelmoussaoui** [comment]
I don't think it is useful to depend on an external crate for this. But yes, having some very basic in-tree URI type that can be used instead of plain `String` would be nice. Do you want to submit a patch for that? There are not that many that I can think of.

Only Account, Screenshot & FileChooser returns a URI.

**Comment by Be-ing** [comment]
This would be a breaking API change. Would you want to make a new major release just for this? Or maybe wait and do this together with other breaking changes?

**Comment by bilelmoussaoui** [comment]
> This would be a breaking API change. Would you want to make a new major release just for this? Or maybe wait and do this together with other breaking changes?

I am planning to make a new major release yes so it is fine to do breaking changes

**Comment by A6GibKm** [comment]
I am not sure this is the best idea. The user can always do
```rust
let uri = url::Url::parse("something")?;
some_ashd_portal(uri.into()).await?;
```
when verification is needed.

**Comment by Be-ing** [comment]
For my use case I am more concerned with properly handling URIs read from ashpd. I was making mistaken assumptions about the returned data which the compiler let me do because it was a String. If its type conveyed it was a URI, that couldn't be ignored.

**Comment by bilelmoussaoui** [comment]
There is no URI inputs, only outputs btw. So we could have just a `pub struct URI(String)` & return that.

**Comment by A6GibKm** [comment]
Doesn't the uri portal require a url input

**Comment by bilelmoussaoui** [comment]
> Doesn't the uri portal require a url input

Hmm yeah indeed.

**Comment by bilelmoussaoui** [comment]
there is also the wallpaper portal that takes a uri

**Comment by A6GibKm** [comment]
I think the best thing to do is to accept arbitrary strings, but return early if, say, url::Url::parse(input) fails.

**Comment by Be-ing** [comment]
I disagree. It would be better to prevent passing invalid URLs by using a URL type as the input.
